Userland/Applications/Dictionary/Dictionary.cpp

```diff
@@ -0,0 +1,198 @@
+    case 's':
+    case 'a':
+        return "adjective";
+    case 'r':
+        return "adverb";
+    default:
+        return "misc";
+    }
+}
+
+static int size_of_index_entry(char const* index_entry)
+{
+    int n = 0;
+    for (int i = 0; i < 16; i++) {
+        if (index_entry[i])
+            n++;
+        else
+            break;
+    }
+    return n;
+}
+
+Dictionary::Dictionary(char const* filename)
+{
+    auto file_or_error = MappedFile::map(filename);
+    if (file_or_error.is_error()) {
+        m_num_words = 0;
+        m_dictionary_bytes = 0;
+        return;
+    }
+
+    m_dictionary_file = file_or_error.value();
+    m_dictionary_bytes = (uint8_t*)m_dictionary_file->data();
+
+    Header const* header = (Header const*)m_dictionary_bytes;
+
+    m_num_words = header->num_words;
+    m_max_word_length = header->max_word_length;
+
+    m_index = offset_to_pointer(header->index_offset);
+    m_word_data_index = offset_to_pointer(header->word_data_index_offset);
+
+    outln("Number of words: {}", m_num_words);
+    outln("Max word length: {}", m_max_word_length);
+
+    VERIFY(m_num_words < 1000 * 1000);
+    VERIFY(m_max_word_length == 16);
+}
+
+StringView Dictionary::word_at_index(int word_index) const
+{
+    if (0 <= word_index && word_index < m_num_words) {
+        char const* c = (char const*)(m_index[word_index]);
+
+        int n = size_of_index_entry(c);
+
+        return StringView(c, n);
+    }
+    return "";
+}
+
+String Dictionary::definition_of(int word_index) const
+{
+    if (!(0 <= word_index && word_index < m_num_words)) {
+        return "";
+    }
+
+    Offset<WordData> word_data_offset = m_word_data_index[word_index];
+
+    WordData const& word_data = *offset_to_pointer(word_data_offset);
+
+    StringBuilder string_builder;
+
+    string_builder.append(word_at_index(word_index));
+    string_builder.append("\n\n");
+
+    char last_part_of_speech = 0;
+    int sense_number = 0;
+
+    int n = word_data.num_senses;
+    for (int i = 0; i < n; i++) {
+        Sense const& sense = word_data.senses[i];
+
+        Definition const& definition = *offset_to_pointer(sense.definition_offset);
+
+        if (definition.part_of_speech != last_part_of_speech) {
+            string_builder.append(part_of_speech_name(definition.part_of_speech));
+            string_builder.append("\n\n");
+
+            last_part_of_speech = definition.part_of_speech;
+            sense_number = 0;
+        }
+
+        if (n > 1)
+            string_builder.append(String::formatted("  {}. ", sense_number + 1));
+
+        string_builder.append(definition.text);
+        string_builder.append("\n\n");
+
+        sense_number += 1;
+    }
+
+    return string_builder.string_view();
+}
+
+void Dictionary::prefix_query(
+    String query,
+    int start_index,
+    int& first_index,
+    int& num_indices) const
+{
+    if (m_num_words == 0) {
+        return;
+    }
+
+    first_index = start_index;
+
+    if (query == "") {
```

Comment:
please check query.is_empty() instead of directly comparting with ""

Reply:
Done.